ostd/src/sync/rcu/non_null.rs

```diff
@@ -0,0 +1,206 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! This module provides a trait and some auxiliary types to help abstract and
+//! work with non-null pointers.
+
+use core::{marker::PhantomData, mem::ManuallyDrop, ops::Deref, ptr::NonNull};
+
+use crate::prelude::*;
+
+/// A trait that abstracts pointers that are non-null.
+///
+/// The most typical examples smart pointer types like `Box<T>` and `Arc<T>`,
+/// which can be converted to and from the raw pointer type of `*const T`.
+///
+/// # Safety
+///
+/// This trait must be implemented correctly (according to the doc comments for
+/// each method). Types like [`Rcu`] rely on this assumption to safely use the
+/// raw pointers.
+///
+/// [`Rcu`]: super::Rcu
+pub unsafe trait NonNullPtr: Send + 'static {
+    /// A type that behaves just like a shared references to the `NonNullPtr`.
+    type Ref<'a>: Deref<Target = Self>
+    where
+        Self: 'a;
+
+    /// Converts to a raw pointer.
+    ///
+    /// Each call to `into_raw` must be paired with a call to `from_raw`
+    /// in order to avoid memory leakage.
+    ///
+    /// The resulting raw pointer must be valid to be immutably accessed
+    /// or borrowed until `from_raw` is called.
```

Comment:
This claim isn't true if we support `Weak<T>`. Please check all the comments and update accordingly.